### ISSUE TITLE
you should apply validation to save the best model

### ISSUE THREAD

**Issue by Levstyle** [minsangkim142/R-net#14]
the structure of your source code is really 6666666! from which, I learned a lot.  while, I have not figured out how to validate the model, when the model instance is created, the variable `is_training` is set, however, it should be set `False` during validation. It cost too much memory to save models, so it's not convenient to save hundreds of models, and validate them one by one after finishing training

**Comment by minsangkim142** [comment]
Hi @Levstyle, what do you mean validation? Do you mean validation of source code to ensure the code is implemented correctly to the paper [here](https://www.microsoft.com/en-us/research/wp-content/uploads/2017/05/r-net.pdf)?
Or do you mean cross-validation of the model to ensure the model is not overfitting?

**Comment by Levstyle** [comment]
cross-validation @minsangkim142 , evaluate the model with validation data after every k batches training, and the best one is kept.

**Comment by minsangkim142** [comment]
Please check the latest commit 5b15ef7ab08cb6d32e5c7fbfbc14de071d267f41, as now we can monitor dev_loss as well as training_loss, dev EM/F1 score in tensorboard. You can control how often to calculate Dev performance from this [line](https://github.com/minsangkim142/R-net/blob/5b15ef7ab08cb6d32e5c7fbfbc14de071d267f41/params.py#L28).

**Comment by Levstyle** [comment]
sorry for my carelessness! the dev data is indeed used to avoid overfitting! to my knowledge, the keep_prob should be less than 1 when training, but it should be 1 when validation. you use the variable is_training to control this, it seems it can not be modified when validation! 

Thanks for your attention !

**Comment by minsangkim142** [comment]
You are totally right about the dropout! The dropout prob is not 0 when calculating dev data and loss. There are two ways to control this,
1. Make variable is_training as a tensorflow variable from tf.placeholder_by_default(True, name="is_train") and feed {model.is_train:False}
2. Make dropout as a tensorflow variable and use the above technique.
What do you think is a better way? Or can you think of any other way of doing this? Thanks for pointing it out.

**Comment by Levstyle** [comment]
I can't agree with you any more. I use the first way in my model, but the structure of my source code is organized badly. Thanks for your awesome code!

**Comment by huangpeng1126** [comment]
Sorry, my question is not related to this issue.
How long does it take to finish 100 epoch training? I have spent 2days (Tesla M40), and current progress is 23th epoch.

**Comment by minsangkim142** [comment]
@huangpeng1126  100 epochs is just an arbitrary number I put, as the network should converge in 10 hours in gtx 1070 (which is around 8 epochs). If you go over that the model is likely to overfit to the training data.

**Comment by JianboTang** [comment]
hi, how about the performance of this model after finishing training?

**Comment by QiXuanWang** [comment]
I've spent 75+ hours on a xeon E5 2.4G machine, average cpu I used is 10+, memory used is 37g, it's now at 73% progress (for 100 epochs I think it's 73 epochs?) and with a F1 number 62%~76%. It's  really slow. Does that mean it's overfit already?

**Comment by minsangkim142** [comment]
Hi @QiXuanWang, are you training in CPU instead of GPU? Then the training might be really slow. Check your tensorboard to monitor your loss to see if your model is overfitting or not.

**Comment by jemmun** [comment]
@minsangkim142 I noted you just spent 10 hours run 8 epochs with gtx1070, it's really faster than mine, I have no idea why i spent about 40 hours with Titanxp just ran 10 epochs. btw: many thanks for your brilliant work.

**Comment by minsangkim142** [comment]
Try changing the number of threads for each of these operations to keep the queue full.
https://github.com/minsangkim142/R-net/blob/74a6abcc0c059d2c3d540947071b20797286c9be/data_load.py#L208-L213
